File: solutions/innerprod.c
/* innerprod.c */
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

#define N 10000000

int main(void)
{
    int *a, *b, ip, i, *localips, tid, localip;
    double t1, t2;

    a = malloc(N * sizeof(*a));
    b = malloc(N * sizeof(*b));
    localips = malloc(omp_get_max_threads()*sizeof(*localips));

    for (i = 0; i < N; i++) {
        a[i] = 2;
        b[i] = 3;
    }

    t1 = omp_get_wtime();
    ip = 0;

#pragma omp parallel private(i,tid,localip) shared (ip,a,b,localips)
    {
      tid = omp_get_thread_num();
      localip = 0;
#pragma omp for nowait
      for (i = 0; i < N; i++)
        localip += a[i] * b[i];
      localips[tid] = localip;
#pragma omp barrier
#pragma omp single
      for (i = 0; i < omp_get_num_threads(); i++)
	ip += localips[i];
    }

    t2 = omp_get_wtime();

    printf("Inner product = %d, time=%g\n", ip, t2-t1);
    return 0;
}

File: solutions/innerprod.f90
PROGRAM innerprod

USE omp_lib
IMPLICIT NONE

INTEGER, PARAMETER :: N = 10000000
INTEGER, ALLOCATABLE :: A(:), B(:), localips(:)
INTEGER :: ip, i, tid, localip
DOUBLE PRECISION :: t1, t2

    ALLOCATE(A(N),B(N),localips(OMP_GET_MAX_THREADS()))
    DO i = 1, N
       A(i) = 2
       B(i) = 3
    ENDDO

    t1 = OMP_GET_WTIME()
    ip = 0
!$OMP PARALLEL PRIVATE(i,tid,localip) SHARED (ip,A,B,localips)
    tid = OMP_GET_THREAD_NUM()
    localip = 0
!$OMP DO
    DO i = 1, N
       localip = localip + A(i) * B(i)
    ENDDO
!$OMP END DO NOWAIT
    localips(tid+1) = localip
!$OMP BARRIER
!$OMP SINGLE
    DO i = 1, OMP_GET_NUM_THREADS()
       ip = ip + localips(i)
    ENDDO
!$OMP END SINGLE
!$OMP END PARALLEL
    t2 = OMP_GET_WTIME()

    WRITE(*,*) 'Inner product = ', ip, 'Time =', t2 - t1

END PROGRAM innerprod
